src/arche/rules/category.py

```diff
@@ -90,17 +91,49 @@ def get_categories(df: pd.DataFrame, max_uniques: int = 10) -> Result:
         the number of unique values less than or equal to `max_uniques` are category columns.
 
     Returns:
-        A result with stats containing value counts.
+        A result with stats containing value counts of categorical columns.
     """
     result = Result("Categories")
 
+    columns = find_likely_cats(df, max_uniques)
     result.stats = [
         value_counts
-        for value_counts in map(lambda c: df[c].value_counts(dropna=False), df)
+        for value_counts in tqdm_notebook(
+            map(lambda c: df[c].value_counts(dropna=False), columns),
+            desc="Finding categories",
+            total=len(columns),
+        )
         if len(value_counts) <= max_uniques
     ]
     if not result.stats:
         result.add_info("Categories were not found")
         return result
     result.add_info(f"{len(result.stats)} category field(s)")
     return result
+
+
+def find_likely_cats(
+    df: pd.DataFrame, max_uniques: int, sample_size: int = 5000
+) -> List[str]:
+    """Find columns which are probably categorical, including nested data.
+    In fact we filter from columns which are certainly not categorical by given `sample_size`.
+    Useful in cases with big datasets and nested data, since `value_counts`
+    performance degrades significantly (100x-10000x) in such cases.
+
+    Args:
+        df: where to find
+        max_uniques: how we decide what is a categorical column
+        sample_size: sample we look in. Defaults to 5000 since for bigger data
+        nested values make `value_counts` really slow
+
+    Returns:
+        List of potential categorical column names.
+    """
+    if len(df) <= sample_size:
+        return df.columns.values
```

Comment:
So, the `max_uniques` named argument has no effect when `df <= sample_size`? That makes the function a little bit inconsistent. I would suggest you to:

- make the calculation consistent by always sampling `df` and filtering by `max_uniques`
- provide a default value for `max_uniques` (an integer or None if you'd like to disable the filter by default)

I'm telling that because as a developer I expect things to change when tweaking function's arguments. Maybe you could just remove this if statement and provide a single path for the function.

Reply:
I thought about it, and made this update, which I think makes everyone happy.
>provide a default value for max_uniques (an integer or None if you'd like to disable the filter by default)

I don't understand why.